src/librustc_mir/interpret/step.rs

```diff
@@ -11,8 +11,20 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
     pub fn inc_step_counter_and_check_limit(&mut self, n: usize) {
         self.terminators_remaining = self.terminators_remaining.saturating_sub(n);
         if self.terminators_remaining == 0 {
-            // FIXME(#49980): make this warning a lint
-            self.tcx.sess.span_warn(self.frame().span, "Constant evaluating a complex constant, this might take some time");
+            let node_id = self
+                    .stack()
+                    .iter()
+                    .rev()
+                    .filter_map(|frame| self.tcx.hir.as_local_node_id(frame.instance.def_id()))
```

Comment:
This should also filter out all frames that are `Allow`, because the lint won't get reported there anyway, but if some frame wants to report the lint, then it should.

This means there might be no frames that we report anything on. In that case just don't emit the lint and only reset the `terminators_remaining` counter.

Reply:
@oli-obk thank you, the suggestions will be addressed soon.